src/librustdoc/html/highlight.rs

```diff
@@ -687,15 +711,30 @@ fn string<T: Display>(
     out: &mut Buffer,
     text: T,
     klass: Option<Class>,
-    context_info: &Option<ContextInfo<'_, '_, '_>>,
+    href_context: &Option<HrefContext<'_, '_, '_>>,
 ) {
+    if let Some(closing_tag) = string_without_closing_tag(out, text, klass, href_context) {
+        out.write_str(closing_tag);
+    }
+}
+
+fn string_without_closing_tag<T: Display>(
+    out: &mut Buffer,
+    text: T,
+    klass: Option<Class>,
+    href_context: &Option<HrefContext<'_, '_, '_>>,
+) -> Option<&'static str> {
```

Comment:
We also use the "optional close tag" pattern in `render_impl` in render/mod.rs.

I think we can make things a little clearer by using the type system:

```
enum ClosingTag {
  A,
  Span,
  Details,
  Div,
  None,
}

impl Display for ClosingTag {
 ...
}
```

Then this function could return `ClosingTag` and in render/mod.rs we could keep track of a `Vec<ClosingTag>`.

Reply:
This function doesn't go "up" to `render/mod.rs`, it's just used in this file as it generates the whole content of the page/code block.

Reply:
Sorry, I misunderstood your comment. I'm not sure if it's a good idea or not. Might be worth to discuss it "outside" of this PR maybe?

Reply:
The contexts and why it's needed are different so having a common mechanism seems a bit weird for me. Not saying it's a bad idea, just that I'm not really convinced it's needed to add a type to handle this.